soroban-env-host/src/storage.rs

```diff
@@ -491,3 +586,157 @@ impl Storage {
+                if can_create_new_objects {
+                    let address_val = self
+                        .add_host_object(cd.contract.metered_clone(self.as_budget())?)?
+                        .into();
+                    res.push(address_val);
+                }
+                match &cd.key {
+                    ScVal::LedgerKeyContractInstance => (),
+                    ScVal::LedgerKeyNonce(n) => {
+                        if can_create_new_objects {
+                            res.push(self.add_host_object(n.nonce)?.into());
+                        }
+                    }
+                    _ => {
+                        if let Some(key) = key_val {
+                            res.push(key);
+                        }
+                    }
+                }
+            }
+            LedgerKey::ContractCode(c) => {
+                if can_create_new_objects {
+                    res.push(
+                        self.add_host_object(self.scbytes_from_hash(&c.hash)?)?
+                            .into(),
+                    );
+                }
+            }
+            LedgerKey::Account(_) | LedgerKey::Trustline(_) => {
+                if can_create_new_objects {
+                    res.push(self.account_address_from_key(lk)?)
+                }
+            }
+            // This shouldn't normally trigger, but it's safer to just return
+            // a safe default instead of an error in case if new key types are
+            // accessed.
+            _ => (),
+        };
+        Ok(res)
+    }
+
+    #[cfg(any(test, feature = "testutils"))]
+    fn check_if_entry_is_live(
+        &self,
+        key: &LedgerKey,
+        entry_with_live_until: &Option<EntryWithLiveUntil>,
+        key_val: Option<Val>,
+    ) -> Result<bool, HostError> {
+        let Some((_, Some(live_until_ledger))) = &entry_with_live_until else {
+            return Ok(true);
+        };
+        let ledger_seq = self
+            .with_ledger_info(|li| Ok(li.sequence_number))
+            .unwrap_or_default();
+        if *live_until_ledger >= ledger_seq {
+            return Ok(true);
+        }
+        match get_key_durability(key) {
+            Some(ContractDataDurability::Temporary) => Ok(false),
+            Some(ContractDataDurability::Persistent) => {
+                let key_type_str = get_key_type_string_for_error(key);
+                let args = self
+                    .as_budget()
+                    .with_observable_shadow_mode(|| self.get_args_for_error(key, key_val, true))
+                    .unwrap_or_else(|_| vec![]);
+                let msg = format!("[testing-only] Accessed {} key that has been archived. Important: this error may only appear in tests; in the real network contracts aren't called at all if any archived entry is accessed.", key_type_str);
+                Err(self.err(
+                    ScErrorType::Storage,
+                    ScErrorCode::InternalError,
```

Comment:
Should this be an internal error if it can be triggered in tests?

Reply:
Yeah, I'm not a fan of this, but the error has to be irrecoverable (in order to imitate the Core logic better) and also it can't be `ExceededLimit`, because that would be even more confusing due to collision with out-of-footprint errors. Ideally we'd have a separate code for this, but that would be pretty hard to implement, given that codes are defined in XDR. So this seemed like the lesser evil.